pkg/cloud/aws/services/userdata/controlplane.go

```diff
@@ -16,111 +16,187 @@ limitations under the License.
-echo -n '{{.SaKey}}' > /etc/kubernetes/pki/sa.key
-chmod 600 /etc/kubernetes/pki/sa.key
-
-PRIVATE_IP=$(curl http://169.254.169.254/latest/meta-data/local-ipv4)
-HOSTNAME="$(curl http://169.254.169.254/latest/meta-data/local-hostname)"
-
-cat >/tmp/kubeadm-controlplane-join-config.yaml <<EOF
----
-apiVersion: kubeadm.k8s.io/v1beta1
-kind: JoinConfiguration
-discovery:
-  bootstrapToken:
-    token: "{{.BootstrapToken}}"
-    apiServerEndpoint: "{{.ELBAddress}}:6443"
-    caCertHashes:
-      - "{{.CACertHash}}"
-nodeRegistration:
-  name: "${HOSTNAME}"
-  criSocket: /var/run/containerd/containerd.sock
-  kubeletExtraArgs:
-    cloud-provider: aws
-controlPlane:
-  localAPIEndpoint:
-    advertiseAddress: "${PRIVATE_IP}"
-    bindPort: 6443
-EOF
-
-kubeadm join --config /tmp/kubeadm-controlplane-join-config.yaml --v 10
+	controlPlaneJoinCloudInit = `{{.Header}}
+write_files:
+-   path: /etc/kubernetes/pki/ca.crt
+    encoding: "base64"
+    owner: root:root
+    permissions: '0640'
+    content: |
+      {{.CACert | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/ca.key
+    encoding: "base64"
+    owner: root:root
+    permissions: '0600'
+    content: |
+      {{.CAKey | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/etcd/ca.crt
+    encoding: "base64"
+    owner: root:root
+    permissions: '0640'
+    content: |
+      {{.EtcdCACert | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/etcd/ca.key
+    encoding: "base64"
+    owner: root:root
+    permissions: '0600'
+    content: |
+      {{.EtcdCAKey | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/front-proxy-ca.crt
+    encoding: "base64"
+    owner: root:root
+    permissions: '0640'
+    content: |
+      {{.FrontProxyCACert | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/front-proxy-ca.key
+    encoding: "base64"
+    owner: root:root
+    permissions: '0600'
+    content: |
+      {{.FrontProxyCAKey | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/sa.pub
+    encoding: "base64"
+    owner: root:root
+    permissions: '0640'
+    content: |
+      {{.SaCert | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/sa.key
+    encoding: "base64"
+    owner: root:root
+    permissions: '0600'
+    content: |
+      {{.SaKey | Base64Encode}}
+
+-   path: /tmp/kubeadm-controlplane-join-config.yaml
+    owner: root:root
+    permissions: '0640'
+    content: |
+      apiVersion: kubeadm.k8s.io/v1beta1
+      kind: JoinConfiguration
+      discovery:
+        bootstrapToken:
+          token: "{{.BootstrapToken}}"
+          apiServerEndpoint: "{{.ELBAddress}}:6443"
+          caCertHashes:
+            - "{{.CACertHash}}"
+      nodeRegistration:
+        name: {{ "{{ ds.meta_data.hostname }}" }}
+        criSocket: /var/run/containerd/containerd.sock
+        kubeletExtraArgs:
+          cloud-provider: aws
+      controlPlane:
+        localAPIEndpoint:
+          advertiseAddress: {{ "{{ ds.meta_data.local_ipv4 }}" }}
+          bindPort: 6443
+kubeadm:
```

Comment:
can we also keep the `--v=10`?

Reply:
To add support, we'll need to extend the cloud-init module here: https://github.com/detiber/cloud-init/tree/cluster-api

Reply:
Is it something that we want in scope of solving #205?

Reply:
I don't think it is in scope for #205 @ashish-amarnath do you mind filing an issue for followup?

Reply:
will do